src/embind/emval.js

```diff
@@ -12,47 +12,36 @@
 /*jslint sub:true*/ /* The symbols 'fromWireType' and 'toWireType' must be accessed via array notation to be closure-safe since craftInvokerFunction crafts functions as strings that can't be closured. */
 
 // -- jshint doesn't understand library syntax, so we need to mark the symbols exposed here
-/*global getStringOrSymbol, emval_handle_array, Emval, __emval_unregister, count_emval_handles, emval_symbols, emval_free_list, get_first_emval, __emval_decref, emval_newers*/
+/*global getStringOrSymbol, emval_handles, Emval, __emval_unregister, count_emval_handles, emval_symbols, __emval_decref, emval_newers*/
 /*global craftEmvalAllocator, emval_addMethodCaller, emval_methodCallers, LibraryManager, mergeInto, emval_allocateDestructors, global, emval_lookupTypes, makeLegalFunctionName*/
 /*global emval_get_global*/
 
 var LibraryEmVal = {
-  $emval_handle_array: [
-    {},
-    {value: undefined},
-    {value: null},
-    {value: true},
-    {value: false}
-  ], // reserve zero and special values
-  $emval_free_list: [],
+  $emval_handles__deps: ['$HandleAllocator'],
+  $emval_handles: "new HandleAllocator();",
   $emval_symbols: {}, // address -> string
 
-  $init_emval__deps: ['$count_emval_handles', '$get_first_emval'],
+  $init_emval__deps: ['$count_emval_handles', '$emval_handles'],
   $init_emval__postset: 'init_emval();',
   $init_emval: function() {
+#if ASSERTIONS
+    assert(emval_handles.allocated.length == 0);
+#endif
+    // reserve zero and special values, these never get de-allocated
+    emval_handles.allocated.push(
+      {},
+      {value: undefined},
+      {value: null},
+      {value: true},
+      {value: false},
+    );
     Module['count_emval_handles'] = count_emval_handles;
-    Module['get_first_emval'] = get_first_emval;
   },
 
-  $count_emval_handles__deps: ['$emval_handle_array'],
+  $count_emval_handles__deps: ['$emval_handles'],
   $count_emval_handles: function() {
-    var count = 0;
-    for (var i = 5; i < emval_handle_array.length; ++i) {
-      if (emval_handle_array[i] !== undefined) {
-        ++count;
-      }
-    }
-    return count;
-  },
-
-  $get_first_emval__deps: ['$emval_handle_array'],
-  $get_first_emval: function() {
-    for (var i = 5; i < emval_handle_array.length; ++i) {
-      if (emval_handle_array[i] !== undefined) {
-        return emval_handle_array[i];
-      }
-    }
-    return null;
+    // Don't include the 5 re-allocated handles
+    return emval_handles.allocated.length - emval_handles.freelist.length - 5;
```

Comment:
Reverted to old code.

Reply:
> since the handle allocator uses a holey array

Uhh maybe it shouldn't? Holey arrays tend to be punished severely performance-wise by optimising engines.

Reply:
Perhaps "holey" is the wrong word here.  Its an array of objects, some of which might be undefined because is set them to undefined.

In any case I believe the old and new code use the array in the exact same way so it shouldn't block this change.

Reply:
> Perhaps "holey" is the wrong word here. Its an array of objects, some of which might be undefined because is set them to undefined.

Ah right but in that case your new code should work I think.

Reply:
Ah yes, good point.  Did you see thomas that I subtracted the number of items in the free list?

Reply:
Actually looks like it does create real holes:

```
delete this.allocated[id];
```

This will create a hole in the array and deoptimize it; it should do `this.allocated[id] = undefined;` or `null` or whatever instead.